lib/fluent/plugin/in_tail.rb

```diff
@@ -392,37 +415,51 @@ def setup_watcher(target_info, pe)
       raise e
     end
 
-    def start_watchers(targets_info)
-      targets_info.each_value { |target_info|
-        pe = nil
-        if @pf
-          pe = @pf[target_info]
-          if @read_from_head && pe.read_inode.zero?
-            begin
-              pe.update(Fluent::FileWrapper.stat(target_info.path).ino, 0)
-            rescue Errno::ENOENT
-              $log.warn "#{target_info.path} not found. Continuing without tailing it."
-            end
+    def construct_watcher(target_info)
+      pe = nil
+      if @pf
+        pe = @pf[target_info]
+        if @read_from_head && pe.read_inode.zero?
+          begin
+            pe.update(Fluent::FileWrapper.stat(target_info.path).ino, 0)
+          rescue Errno::ENOENT
+            $log.warn "#{target_info.path} not found. Continuing without tailing it."
           end
         end
+      end
 
-        begin
-          tw = setup_watcher(target_info, pe)
-        rescue WatcherSetupError => e
-          log.warn "Skip #{target_info.path} because unexpected setup error happens: #{e}"
-          next
-        end
+      begin
+        tw = setup_watcher(target_info, pe)
+      rescue WatcherSetupError => e
+        log.warn "Skip #{target_info.path} because unexpected setup error happens: #{e}"
+        return
+      end
 
-        begin
-          target_info = TargetInfo.new(target_info.path, Fluent::FileWrapper.stat(target_info.path).ino)
-          @tails[target_info] = tw
-        rescue Errno::ENOENT
-          $log.warn "stat() for #{target_info.path} failed with ENOENT. Drop tail watcher for now."
-          # explicitly detach and unwatch watcher `tw`.
-          tw.unwatched = true
-          detach_watcher(tw, target_info.ino, false)
+      begin
+        target_info = TargetInfo.new(target_info.path, Fluent::FileWrapper.stat(target_info.path).ino)
+        @tails[target_info] = tw
+      rescue Errno::ENOENT
+        $log.warn "stat() for #{target_info.path} failed with ENOENT. Drop tail watcher for now."
+        # explicitly detach and unwatch watcher `tw`.
+        tw.unwatched = true
+        detach_watcher(tw, target_info.ino, false)
+      end
+    end
+
+    def start_watchers(targets_info)
+      if @enable_read_bytes_limit_thread_pool
+        @thread_pool = TailThread::Pool.new(@max_thread_pool_size) do |pool|
+          targets_info.each_value {|target_info|
+            pool.run {
+              construct_watcher(target_info)
```

Comment:
Probably only start-up task is multi-threaded, trailing all timer or inotify handlers are run on the same (main) thread because they are invoked by the main event loop.

Reply:
Hmm..., that's right.
Using multi-threaded only for start-up tasks is not enough for preventing to block from stop the world with `#sleep`.
But, I guess that we needn't to use queue which is constructed by thread pool.

Reply:
Fixed in https://github.com/fluent/fluentd/pull/3185/commits/478b487383e1dec1af86cb0def2000b515eb6577.